apis/rabbitmq/v1beta1/rabbitmq_webhook.go

```diff
@@ -0,0 +1,175 @@
+func (r *RabbitMq) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/mutate-rabbitmq-openstack-org-v1beta1-rabbitmq,mutating=true,failurePolicy=fail,sideEffects=None,groups=rabbitmq.openstack.org,resources=rabbitmqs,verbs=create;update,versions=v1beta1,name=mrabbitmq.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Defaulter = &RabbitMq{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *RabbitMq) Default() {
+	rabbitmqlog.Info("default", "name", r.Name)
+
+	r.Spec.Default()
+}
+
+// Default - set defaults for this RabbitMq spec
+func (spec *RabbitMqSpec) Default() {
+	if spec.ContainerImage == "" {
+		spec.ContainerImage = rabbitMqDefaults.ContainerImageURL
+	}
+	spec.RabbitMqSpecCore.Default()
+}
+
+// Default - common validations go here (for the OpenStackControlplane which uses this one)
+func (spec *RabbitMqSpecCore) Default() {
+	//nothing to validate yet
+}
+
+// TODO(user): change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
+//+kubebuilder:webhook:path=/validate-rabbitmq-openstack-org-v1beta1-rabbitmq,mutating=false,failurePolicy=fail,sideEffects=None,groups=rabbitmq.openstack.org,resources=rabbitmqs,verbs=create;update,versions=v1beta1,name=vrabbitmq.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &RabbitMq{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *RabbitMq) ValidateCreate() (admission.Warnings, error) {
+	rabbitmqlog.Info("validate create", "name", r.Name)
+	var allErrs field.ErrorList
+	var allWarn []string
+	basePath := field.NewPath("spec")
+
+	// When a TopologyRef CR is referenced, fail if a different Namespace is
+	// referenced because is not supported
+	if r.Spec.TopologyRef != nil {
+		if err := topologyv1.ValidateTopologyNamespace(r.Spec.TopologyRef.Namespace, *basePath, r.Namespace); err != nil {
+			allErrs = append(allErrs, err)
+		}
+	}
+	allErrs = append(allErrs, common_webhook.ValidateDNS1123Label(
+		field.NewPath("metadata").Child("name"),
+		[]string{r.Name},
+		CrMaxLengthCorrection,
+	)...) // omit issue with  statefulset pod label "controller-revision-hash": "<statefulset_name>-<hash>"
+
+	if len(allErrs) != 0 {
+		return allWarn, apierrors.NewInvalid(
+			schema.GroupKind{Group: "rabbitmq.openstack.org", Kind: "RabbitMq"},
+			r.Name, allErrs,
+		)
+	}
+
+	return allWarn, nil
+
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *RabbitMq) ValidateUpdate(_ runtime.Object) (admission.Warnings, error) {
+	rabbitmqlog.Info("validate update", "name", r.Name)
+
+	var allErrs field.ErrorList
+	var allWarn []string
+	basePath := field.NewPath("spec")
+
+	// When a TopologyRef CR is referenced, fail if a different Namespace is
+	// referenced because is not supported
+	if r.Spec.TopologyRef != nil {
```

Comment:
(for a follow up) Wondering if we should have a single SpecCore ValidateCreate and ValidateUpdate as part of the `RabbitMqSpec` that returns the SpecCore related function, so we only have to write it once. I saw this pattern in many operators, maybe we can think about it for a follow up.